File: locations/break_room.pl
:- multifile describe/1, go/1,
grape_soda/0, strange_soda/0.

describe(break_room) :-
    write("You are in the break room"), nl,
    write("There's a vending machine here (HINT: machine.)"), nl,
    write("You can go back to the kitchen"), !.

machine :-
    (not(grape_soda), not(strange_soda) ->
        machine_keyboard, !
    ;
        write("(You've already used the coupon)"), nl
    ).

machine_keyboard :-
    nl,
    write("Enter a number:"), nl,
    read(Number),
    (not(integer(Number)) ->
        nl,
        write("(Not a valid input)"), nl
    ;
        true
    ),

    (member(Number, [034, 015, 092, 059, 041, 065, 026, 087]) ->
        nl,
        write("Can of some strange soda that you've never heard of rolled out... (it's not Grape Soda but it's better than nothing)"), nl,
        assert(strange_soda),
        assert(holding(strange_soda)), !
    ;
        true
    ),

    (not(strange_soda) ->
        (Number == 073 ->
            nl,
            write("You have the Grape Soda!"), nl,
            assert(grape_soda),
            assert(holding(grape_soda))
        ;
            nl,
            write("Nothing happens..."), nl
        )
    ;
        true
    ).

File: Haskell/BreakRoom.hs
module BreakRoom where

import Game

describe :: IO ()
describe = do
  putStrLn "You are in the break room"
  putStrLn "There's a vending machine here (HINT: machine)"
  putStrLn "You can go back to the kitchen from here"

machine :: Game -> IO Game
machine (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory) = do
  if (creedQuest == CreedQuestInProgress) then do
    newGame <- machineKeyboard (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory)
    return newGame
  else if (creedQuest == CreedQuestNotStarted) then do
    putStrLn ""
    putStrLn "(It might be useful later)"
    putStrLn ""
    return (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory)
  else do
    putStrLn ""
    putStrLn "(You've already used the coupon)"
    putStrLn ""
    return (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory)

isNumber :: String -> Bool
isNumber str =
  case reads str :: [(Double, String)] of
    [(_, "")] -> True
    _         -> False

machineKeyboard :: Game -> IO Game
machineKeyboard (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory) = do
  putStrLn ""
  putStrLn "Enter a number:"
  input <- getLine
  if (not (isNumber input)) then do
    putStrLn ""
    putStrLn "(Not a valid input)"
    putStrLn ""
    return (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory)
  else if (input `elem` ["034", "015", "092", "059", "041", "065", "026", "087"]) then do
    putStrLn ""
    putStrLn "Can of some strange soda that you've never heard of rolled out... (it's not Grape Soda but it's better than nothing)"
    putStrLn ""
    let newInventory = inventory ++ [StrangeSodaItem]
    return (Game iamAt saidHi jimQuest StrangeSoda dwightQuest newInventory)
  else do
    if input == "073"
      then do
        putStrLn ""
        putStrLn "You have the Grape Soda!"
        putStrLn ""
        let newInventory = inventory ++ [GrapeSodaItem]
        return (Game iamAt saidHi jimQuest GrapeSoda dwightQuest newInventory)
      else do
        putStrLn ""
        putStrLn "Nothing happens..."
        putStrLn ""
        return (Game iamAt saidHi jimQuest creedQuest dwightQuest inventory)
